orttraining/orttraining/include/runner/orttraining_c_api.h

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) SignalPop LLC. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include <string.h>
+#include <runner\training_runner.h>
+
+using namespace onnxruntime;
+using namespace onnxruntime::common;
+using namespace std;
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum OrtTrainingOptimizer {
+    ORT_TRAINING_OPTIMIZER_SGD = 0
+};
+
+typedef enum OrtTrainingLossFunction {
+    ORT_TRAINING_LOSS_FUNCTION_SOFTMAXCROSSENTROPY = 0
+};
+
+typedef enum OrtTrainingStringParameter {
+    ORT_TRAINING_MODEL_PATH = 0,
+    ORT_TRAINING_LOG_PATH = 1,
+    ORT_TRAINING_INPUT_LABELS = 2,
+    ORT_TRAINING_OUTPUT_PREDICTIONS = 3,
+    ORT_TRAINING_OUTPUT_LOSS = 4
+};
+
+typedef enum OrtTrainingLongParameter {
+  ORT_TRAINING_NUM_TRAIN_STEPS = 0,
+  ORT_TRAINING_TRAIN_BATCH_SIZE = 1,
+  ORT_TRAINING_EVAL_BATCH_SIZE = 2,
+  ORT_TRAINING_EVAL_PERIOD = 3,
+  ORT_TRAINING_DISPLAY_LOSS_STEPS = 4
+};
+
+typedef enum OrtTrainingNumericParameter {
+    ORT_TRAINING_LEARNING_RATE = 0,
+};
+
+typedef enum OrtTrainingBooleanParameter {
+  ORT_TRAINING_USE_GIST = 0,
+  ORT_TRAINING_USE_CUDA = 1,
+  ORT_TRAINING_USE_PROFILER = 2,
+  ORT_TRAINING_USE_TENSORBOARD = 3,
+  ORT_TRAINING_IS_PERFTEST = 4,
+  ORT_TRAINING_SHUFFLE_DATA = 5
+};
+
+typedef enum OrtDataUse {
+  ORT_DATAUSE_TRAINING = 0,
+  ORT_DATAUSE_TESTING = 1
+};
+
+struct OrtValueCollection;
+
+typedef void(__stdcall* OrtErrorFunctionCallback)(_In_ const size_t ulCount, _In_ OrtValueCollection* output);  // Array of outputs (e.g. label, predictions, loss).
+typedef void(__stdcall* OrtEvaluationFunctionCallback)(size_t num_samples, size_t step);
+typedef void(__stdcall* OrtDataGetBatchCallback)(_In_ const size_t ulBatchSize, _In_ const size_t ulCount, _In_ OrtValueCollection* data); // Array of input locations, then set by call-back implemenation by calling SetAt.
+
+
+// The actual types defined have an Ort prefix
+//ORT_RUNTIME_CLASS(OrtValueCollection);
+struct OrtValueCollection {
+  std::vector<OrtValue*> m_rgValues;
```

Comment:
It looks like this is C++ and uses the STL in a public C API header file. Unfortunately even if it was a C++ header file you can't use STL types in public headers since there is no guarantee they'll be binary compatible.

Is there any way to extend an existing type? Like OrtValue and add a new type to it? (like how it can be a tensor/set/map already). It must be an opaque type from the outside similar to OrtValue.

Reply:
@RyanUnderhill ,
I don't see why not.  Since the capacity is fixed at construction it would be very easy to just use an array of OrtValue pointers.  I was unaware of the STL limitations and only used these types for convenience but they are not required.  I will re-work this to use an array of OrtValue pointers for the next update.

Reply:
@RyanUnderhill ,
For the C header files, are there any restrictions from structures defining constructors and destructors?

Reply:
@RyanUnderhill ,
I can get the header down to the following with no constructors, destructors or any header file includes:
struct OrtValueCollection {
  size_t capacity_;
  size_t count_;
  OrtValue** rgvalues_;
  char** rgnames_;
  bool* rgname_owned_;
};

struct OrtTrainingApi;
typedef struct OrtTrainingApi OrtTrainingApi;

struct InternalParameters;
typedef struct InternalParameters InternalTrainingParameters;

struct onnxruntime::training::TrainingRunner::Parameters;

// The actual types defined have an Ort prefix
//ORT_RUNTIME_CLASS(TrainingParameters);
struct OrtTrainingParameters {
  struct onnxruntime::training::TrainingRunner::Parameters* prunner_param_;
  InternalTrainingParameters* pinternal_param_;
  char* pszinput_labels_;
  char* pszoutput_predictions_;
  char* pszoutput_loss_;
  char* pszloss_function_type_;
  char* pszinit_feed_names_;
  bool use_cuda_;
  bool use_tensorboard_;
  OrtDataGetBatchCallback fntraining_data_getbatch_;
  OrtDataGetBatchCallback fntesting_data_getbatch_;
  OrtValueCollection* ptraining_data_;
  OrtValueCollection* ptesting_data_;
  onnxruntime::training::TrainingRunner* ptraining_runner_;
  onnxruntime::training::IDataLoader* ptraining_data_loader_;
  onnxruntime::training::IDataLoader* ptesting_data_loader_;
};
typedef struct OrtTrainingParameters OrtTrainingParameters;

Still working on fixing a couple of issues, but when you all are ready for me to update, just let me know.

Reply:
The key to any of these is to make any structrure or class opaque for the user and move all the definitions into .cc

Reply:
Yes, as any type defined in the header can never be changed since the user's code sees it. An opaque type can be implemented in any way internally. Typically only structure the user can also implement are in the header. Otherwise it's an opaque type.

Also, with an opaque type you can use C++ types internally. You just need to exports all functions that manipulate the type, just like the existing C API does.

Reply:
@RyanUnderhill, Agreed.  I will make these more opaque before the next update.  Where do you draw the line though?  Given that all of the items in the struct are only manipulated by the functions themselves, I could just do a forward struct definition, such as "struct OrtTrainingParameters", and then move the full structure definition to the CC file.  Same would work for the OrtValueCollection type.  That would give much more freedom for future changes without impacting the user code and is easy to do.

Reply:
@yuslepukhin , @RyanUnderhill 
Good suggestions.  I now have just the forward definitions of the structs in the header as:

struct OrtValueCollection;
struct OrtTrainingParameters;

Much cleaner.  And the funny thing is that I could probably have kept the original STL code now that it is all in the CC file.  I have already taken that out, so I'll leave it as it is for now.